ast/src/asm_analysis/display.rs

```diff
@@ -1,77 +1,81 @@
 }
 
+/// quick and dirty String to String indentation
+fn indent<S: ToString>(s: S, indentation: usize) -> String {
+    s.to_string()
+        .split('\n')
+        .map(|line| format!("{}{line}", "\t".repeat(indentation)))
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+/// quick and dirty String to String way to comment out
+fn comment_out<S: ToString>(s: S) -> String {
+    s.to_string()
+        .split('\n')
+        .map(|line| format!("//{line}"))
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
 impl<T: Display> Display for Machine<T> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match (&self.latch, &self.function_id) {
+            (Some(latch), Some(function_id)) => write!(f, "({latch}, {function_id})"),
+            (None, None) => write!(f, ""),
+            (Some(latch), None) => write!(f, "({latch}, _)"),
+            (None, Some(function_id)) => write!(f, "(_, {function_id})"),
+        }?;
+
+        writeln!(f, " {{")?;
+
         // TODO: implement indentation properly (passing a context to the visitor)
         for s in &self.degree {
-            writeln!(f, "\t{s}")?;
+            writeln!(f, "{}", indent(s, 1))?;
         }
         for s in &self.registers {
-            writeln!(f, "\t{s}")?;
+            writeln!(f, "{}", indent(s, 1))?;
         }
         for s in &self.constraints {
-            writeln!(f, "\t{s}")?;
+            writeln!(f, "{}", indent(s, 1))?;
         }
         for i in &self.instructions {
-            writeln!(f, "\t{i}")?;
+            writeln!(f, "{}", indent(i, 1))?;
         }
         for o in &self.functions {
-            writeln!(f, "\t{o}")?;
+            writeln!(f, "{}", indent(o, 1))?;
         }
         if let Some(rom) = &self.rom {
-            writeln!(f, "{rom}")?;
+            writeln!(f, "{}", indent(comment_out(rom), 1),)?;
         }
-        Ok(())
+        writeln!(f, "}}")
     }
 }
 
 impl<T: Display> Display for Rom<T> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        writeln!(f, "\t// rom {{")?;
-        let statements = &mut self.statements.iter();
-        match &self.batches {
-            Some(batches) => {
-                for batch in batches {
-                    for statement in statements.take(batch.size) {
-                        writeln!(f, "\t// \t{}", statement)?;
-                    }
-                    writeln!(
-                        f,
-                        "\t// \t// END BATCH{}",
-                        batch
-                            .reason
-                            .as_ref()
-                            .map(|reason| format!(" {reason}"))
-                            .unwrap_or_default()
-                    )?;
-                }
-            }
-            None => {
-                for statement in statements {
-                    writeln!(f, "\t// \t {}", statement)?;
-                }
-            }
-        }
-        write!(f, "\t// }}")
+        writeln!(f, "rom {{")?;
```

Comment:
This was commented out before, why not anymore?

Reply:
It just moved to the caller
```rust
writeln!(f, "{}", indent(comment_out(rom), 1),)?;
```